engine/engine_test.go

```diff
@@ -2901,6 +2923,131 @@ func TestEngineRecoversFromPanic(t *testing.T) {
 
 }
 
+func TestSparseHistogram(t *testing.T) {
+	opts := promql.EngineOpts{
+		Timeout:              1 * time.Hour,
+		MaxSamples:           1e10,
+		EnableNegativeOffset: true,
+		EnableAtModifier:     true,
+	}
+
+	cases := []struct {
+		name        string
+		count       int
+		metricsName string
+		labels      []string
+		query       string
+		mixedSample bool
+	}{
+		{
+			name:        "histogram_quantile(.80) with sparse histogram and old histograms",
+			count:       100,
+			metricsName: "sparse_histogram_series",
+			labels:      []string{"test", "a", "test2", "b"},
+			query:       "histogram_quantile(.80,rate(sparse_histogram_series[1m]))",
+			mixedSample: true,
+		},
+		{
+			name:        "histogram_sum() with sparse histogram and old histograms",
+			count:       200,
+			metricsName: "sparse_histogram_series",
+			labels:      []string{"test", "a", "test2", "b"},
+			query:       "histogram_sum(rate(sparse_histogram_series[1m]))",
+			mixedSample: true,
+		},
+		{
+			name:        "histogram_sum() with sparse histogram",
+			count:       200,
+			metricsName: "sparse_histogram_series",
+			labels:      []string{"test", "a", "test2", "b"},
+			query:       "histogram_sum(rate(sparse_histogram_series[1m]))",
+		},
+		{
+			name:        "histogram_count() with sparse histogram",
+			count:       120,
+			metricsName: "sparse_histogram_series",
+			labels:      []string{"test", "a", "test2", "b"},
+			query:       "histogram_count(rate(sparse_histogram_series[1m]))",
+		},
+		{
+			name:        "histogram_quantile(.80) with sparse histogram",
+			count:       100,
+			metricsName: "sparse_histogram_series",
+			labels:      []string{"test", "a", "test2", "b"},
+			query:       "histogram_quantile(.80,rate(sparse_histogram_series[1m]))",
+		},
+		{
+			name:        "rate() with sparse histogram",
+			count:       100,
+			metricsName: "sparse_histogram_series",
+			query:       "rate(sparse_histogram_series[1m])",
+		},
+		{
+			name:        "increase() with sparse histogram",
+			count:       100,
+			metricsName: "sparse_histogram_series",
+			query:       "increase(sparse_histogram_series[1m])",
+		},
+		{
+			name:        "delta() with sparse histogram",
+			count:       100,
+			metricsName: "sparse_histogram_series",
+			query:       "delta(sparse_histogram_series[1m])",
+		},
```

Comment:
We should add a test for the calculating `rate` on both native histograms and single counter metrics.